app/helpers/admin/audit_trail_helper.rb

```diff
@@ -27,20 +27,32 @@ def paginated_audit_trail_url(page)
   end
 
   def render_editorial_remarks(remarks, edition)
-    this_edition_remarks, other_edition_remarks = remarks.partition { |r| r.edition == edition }
+    previous_state = nil
     out = ""
-    if this_edition_remarks.any?
-      out << tag.h2("On this edition")
-      out << tag.ul(class: "list-unstyled") do
-        render partial: "admin/editions/audit_trail_entry", collection: this_edition_remarks
-      end
-    end
-    if other_edition_remarks.any?
-      out << tag.h2("On previous editions", class: "add-top-margin")
-      out << tag.ul(class: "list-unstyled") do
-        render partial: "admin/editions/audit_trail_entry", collection: other_edition_remarks
+    list = ""
+    remarks.query.each do |remark|
+      current_state = if remark.edition_id > edition.id
+                        :newer
+                      elsif remark.edition_id < edition.id
+                        :previous
+                      else
+                        :current
+                      end
+
+      if current_state != previous_state
+        out << tag.ul(class: "list-unstyled") { list.html_safe } unless list.empty?
+        list = ""
+        out << case current_state
+               when :newer then tag.h2("On newer editions", class: "add-top-margin")
+               when :previous then tag.h2("On previous editions", class: "add-top-margin")
+               when :current then tag.h2("On this edition", class: "add-top-margin")
+               end
       end
+
+      list << render(partial: "admin/editions/remark_entry", locals: { remark: remark })
+      previous_state = current_state
     end
+    out << tag.ul(class: "list-unstyled") { list.html_safe }
     out.html_safe
```

Comment:
I wasn't sure whether to (and how best to) refactor this. Any thoughts greatly appreciated.

Reply:
Hmm yes, does this look quite a complicated method and can understand it could benefit from a refactoring. 

I'm also a bit unsure of how best to do it though as it's quite interconnected. I've had a quick go at breaking it up into two methods with separate concerns, one to group the items and the other to turn a group into HTML.

```
  def render_editorial_remarks(remarks, edition)
    previous_state = nil
    groups = remarks.query.reduce([]) do |memo, remark|
      current_state = if remark.edition_id > edition.id
                        :newer
                      elsif remark.edition_id < edition.id
                        :previous
                      else
                        :current
                      end
      memo << { heading_type: current_state, remarks: [] } if current_state != previous_state
      memo.last[:remarks] << remark
    end
    rendered_groups = groups.map { |g| render_editorial_remarks_list(g[:remarks], g[:heading_type]) })
    safe_join(rendered_groups)
  end

  def render_editorial_remarks_list(remarks, heading_type)
    out = ""
    out << case current_state
           when :newer then tag.h2("On newer editions", class: "add-top-margin")
           when :previous then tag.h2("On previous editions", class: "add-top-margin")
           when :current then tag.h2("On this edition", class: "add-top-margin")
           end

    remark_items = remarks.map { |r| render(partial: "admin/editions/remark_entry", locals: { remark: r }) }
    out << tag.ul(class: "list-unstyled") { safe_join(remark_items) }
    out.html_safe
  end
```

Do you know if there are tests for this helper? it feels like we should probably add some if not.

Reply:
I've taken a stab at refactoring this, too. I've squashed my changes in on this branch, but I've also kept them separate over here – 2435981d427ced2d2f1362249530bf70cdea3d8a – so that @jyoung-gds can follow what I changed when he returns from leave.